cmd/collectors/zapi/plugins/shelf/shelf.go

```diff
@@ -139,51 +144,100 @@ func (my *Shelf) Init() error {
 	}
 
 	my.Logger.Debug().Msgf("initialized with data [%d] objects", len(my.data))
+
+	// setup batchSize for request
+	if my.client.IsClustered() {
+		if b := my.Params.GetChildContentS("batch_size"); b != "" {
+			if _, err := strconv.Atoi(b); err == nil {
+				my.batchSize = b
+			}
+		} else {
+			my.batchSize = BatchSize
+		}
+	}
 	return nil
 }
 
 func (my *Shelf) Run(data *matrix.Matrix) ([]*matrix.Matrix, error) {
 
 	var (
-		result *node.Node
-		err    error
+		result     *node.Node
+		ad, pd     time.Duration // Request/API time, Parse time, Fetch time
+		output     []*matrix.Matrix
+		shelves    int
+		numMetrics int
+		err        error
 	)
 
+	apiT := 0 * time.Second
+	parseT := 0 * time.Second
+
 	if !my.client.IsClustered() {
 		for _, instance := range data.GetInstances() {
 			instance.SetLabel("shelf", instance.GetLabel("shelf_id"))
 		}
 	}
 
-	if result, err = my.client.InvokeRequestString(my.query); err != nil {
-		return nil, err
-	}
-
 	// Set all global labels from zapi.go if already not exist
 	for a := range my.instanceLabels {
 		my.data[a].SetGlobalLabels(data.GetGlobalLabels())
 	}
 
-	var output []*matrix.Matrix
+	request := node.NewXMLS(my.query)
+	tag := "initial"
+	totalShelves := 0
+	totalMetrics := 0
 
-	if my.client.IsClustered() {
-		output, err = my.handleCMode(result)
-	} else {
-		output, err = my.handle7Mode(result)
-	}
-	if err != nil {
-		return output, err
+	if my.client.IsClustered() && my.batchSize != "" {
+		request.NewChildS("max-records", my.batchSize)
 	}
 
-	if my.client.IsClustered() {
-		return my.calculateEnvironmentMetrics(output, data)
-	} else {
-		return output, nil
+	for {
+		if result, tag, ad, pd, err = my.client.InvokeBatchWithTimers(request, tag); err != nil {
+			return nil, err
+		}
+
+		if result == nil {
+			break
+		}
+
+		apiT += ad
+		parseT += pd
+
+		var batchOutput []*matrix.Matrix
+
+		if my.client.IsClustered() {
+			batchOutput, shelves, numMetrics, err = my.handleCMode(result)
+		} else {
+			batchOutput, shelves, numMetrics, err = my.handle7Mode(result)
+		}
+
+		if err != nil {
+			return output, err
+		}
+
+		if my.client.IsClustered() {
+			if err = my.calculateEnvironmentMetrics(data); err != nil {
+				return output, err
+			}
+		}
+		totalShelves += shelves
+		totalMetrics += numMetrics
+		output = append(output, batchOutput...)
 	}
 
+	my.Logger.Info().
```

Comment:
I think this is redundant with the recently added collector level logging. With this line, are you seeing double logging? If so, let's remove this one unless we think it's providing more value than the collector-level one.

Reply:
not required, removed.